File: 77. 组合/swift/main/main/main.swift
//
//  main.swift
//  main
//
//  Created by ByteDance on 4/28/25.
//

import Foundation

class Solution {
    var result = [[Int]]()
    var path = [Int]()

    func combine(_ n: Int, _ k: Int) -> [[Int]] {
        backtracking(n: n, k: k, startIndex: 1)

        return result
    }

    func backtracking(n: Int, k: Int, startIndex: Int) {
        if path.count == k {
            result.append(path)
            return
        }
        guard startIndex <= n else { return }
        for i in startIndex ... n {
            path.append(i)
            backtracking(n: n, k: k, startIndex: i + 1)
            _ = path.popLast()
        }
    }
}

let s = Solution()
print(s.combine(4, 2))

File: 77. 组合/rust/main/src/main.rs
struct Solution {}

impl Solution {
    pub fn combine(n: i32, k: i32) -> Vec<Vec<i32>> {
        let mut result: Vec<Vec<i32>> = vec![];
        let mut path: Vec<i32> = vec![];

        Self::backtracking(&mut result, &mut path, n, k, 1);

        result
    }

    fn backtracking(
        result: &mut Vec<Vec<i32>>, 
        path: &mut Vec<i32>, 
        n: i32, 
        k: i32,
        start_index: i32
    ) {
        if path.len() == k as usize {
            result.push(path.to_vec());
            return
        }
        for i in start_index ..= n {
            path.push(i);
            Self::backtracking(result, path, n, k, i + 1);
            path.pop();
        }
    }
}

fn main() {
    let s = Solution{};
    println!("{:?}", Solution::combine(4, 2));
}
